pkg/controllers/jobset_controller.go

```diff
@@ -950,9 +884,159 @@ func findJobFailureTime(job *batchv1.Job) *metav1.Time {
+}
+
+// enqueueEvent appends a new k8s event to be emitted if and only after running the status
+// update functions in the updateStatusOpts, the status update API call suceeds.
+func enqueueEvent(updateStatusOpts *statusUpdateOpts, event *eventParams) {
+	updateStatusOpts.events = append(updateStatusOpts.events, event)
+}
+
+func setJobSetCompleted(js *jobset.JobSet, updateStatusOpts *statusUpdateOpts) {
+	setCondition(&conditionOpts{
+		jobset:    js,
+		eventType: corev1.EventTypeNormal,
+		condition: metav1.Condition{
+			Type:    string(jobset.JobSetCompleted),
+			Status:  metav1.ConditionStatus(corev1.ConditionTrue),
+			Reason:  constants.AllJobsCompletedReason,
+			Message: constants.AllJobsCompletedMessage,
+		},
+	}, updateStatusOpts)
+}
+
+func setJobSetFailed(ctx context.Context, js *jobset.JobSet, reason, msg string, updateStatusOpts *statusUpdateOpts) {
+	setCondition(&conditionOpts{
+		jobset: js,
+		condition: metav1.Condition{
+			Type:    string(jobset.JobSetFailed),
+			Status:  metav1.ConditionStatus(corev1.ConditionTrue),
+			Reason:  reason,
+			Message: msg,
+		},
+		eventType: corev1.EventTypeWarning,
+	}, updateStatusOpts)
+}
+
+func setInOrderStartupPolicyInProgress(js *jobset.JobSet, updateStatusOpts *statusUpdateOpts) {
+	// Add a condition to the JobSet indicating the in order startup policy is executing.
+	setCondition(&conditionOpts{
+		jobset:           js,
+		eventType:        corev1.EventTypeNormal,
+		forceFalseUpdate: true,
+		condition:        inOrderStartupPolicyInProgressCondition(),
+	}, updateStatusOpts)
+}
+
+func setInOrderStartupPolicyCompleted(js *jobset.JobSet, updateStatusOpts *statusUpdateOpts) {
+	setCondition(&conditionOpts{
+		jobset:    js,
+		eventType: corev1.EventTypeNormal,
+		condition: inOrderStartupPolicyCompletedCondition(),
+	}, updateStatusOpts)
+}
+
+func setJobSetSuspended(js *jobset.JobSet, updateStatusOpts *statusUpdateOpts) {
+	setCondition(&conditionOpts{
+		jobset:    js,
+		eventType: corev1.EventTypeNormal,
+		condition: metav1.Condition{
+			Type:               string(jobset.JobSetSuspended),
+			Status:             metav1.ConditionStatus(corev1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+			Reason:             constants.JobSetSuspendedReason,
+			Message:            constants.JobSetSuspendedMessage,
+		},
+	}, updateStatusOpts)
+}
+
+func setJobSetResumed(js *jobset.JobSet, updateStatusOpts *statusUpdateOpts) {
+	setCondition(&conditionOpts{
+		jobset:    js,
+		eventType: corev1.EventTypeNormal,
+		condition: metav1.Condition{
+			Type:               string(jobset.JobSetSuspended),
+			Status:             metav1.ConditionStatus(corev1.ConditionFalse),
+			LastTransitionTime: metav1.Now(),
+			Reason:             constants.JobSetResumedReason,
+			Message:            constants.JobSetResumedMessage,
+		},
+	}, updateStatusOpts)
+}
+
+// function parameters for setCondition
+type conditionOpts struct {
+	jobset *jobset.JobSet
```

Comment:
we don't need to carry a reference to jobset, we will not be updating the status of more than one jobset in each reconcile and the reference should be available at the time we call whatever function that consumes conditionOpts.

Reply:
Moved jobset reference out of conditionOpts into its own function parameter in `setCondition()`, to match the pattern in `updateCondition()`.